lib/Optimizer/Transforms/LoopPeeling.cpp

```diff
@@ -0,0 +1,125 @@
+    LLVM_DEBUG(llvm::dbgs() << "will peel do-while loop:\n"; loop.dump());
+
+    // Split the block after loop, old-loop-block and after-block.
+    Operation *nextOp = loop->getNextNode();
+    auto *oldLoopBlock = rewriter.getInsertionBlock();
+    auto loopPos = rewriter.getInsertionPoint();
+    auto *afterBlock =
+        rewriter.splitBlock(oldLoopBlock, Block::iterator{nextOp});
+
+    // Add terminator to old-loop-block: thread results of loop to after-block's
+    // arguments and replace the loop results uses.
+    for (auto res : loop.getResults())
+      afterBlock->addArgument(res.getType(), loop.getLoc());
+    rewriter.setInsertionPointToEnd(oldLoopBlock);
+    auto finalBranch = rewriter.create<cf::BranchOp>(loop.getLoc(), afterBlock,
+                                                     loop.getResults());
+    // NB: the results of the original loop are now split between the peeled
+    // copy of body and the modified new loop. Introduce explicit block
+    // arguments for the phi node functionality.
+    for (auto iter : llvm::enumerate(loop.getResults())) {
+      Value v = iter.value();
+      v.replaceUsesWithIf(
+          finalBranch.getDestOperands()[iter.index()],
+          [branch = finalBranch.getOperation()](OpOperand &opnd) {
+            auto *op = opnd.getOwner();
+            return op != branch;
+          });
+    }
+
+    // Split the block before loop, before-block and new-loop-block. Add branch
+    // arguments to new-loop-block corresponding to loop's region arguments.
+    auto *beforeBlock = oldLoopBlock;
+    auto *newLoopBlock = rewriter.splitBlock(oldLoopBlock, loopPos);
+    for (auto res : loop.getResults())
+      newLoopBlock->addArgument(res.getType(), loop.getLoc());
+    SmallVector<Value> loopArgs = loop.getOperands();
+    loop.getInitialArgsMutable().assign(newLoopBlock->getArguments());
+
+    // Clone the body region. Add branch from before-loop to the entry block of
+    // the cloned CFG.
+    rewriter.cloneRegionBefore(loop.getBodyRegion(), newLoopBlock);
+    Block *firstBlock = beforeBlock->getNextNode();
+    rewriter.setInsertionPointToEnd(beforeBlock);
+    rewriter.create<cf::BranchOp>(loop.getLoc(), firstBlock, loopArgs);
+
+    // Replace continue ops with branches to the new-loop-block. Replace break
+    // ops with branches to the after-block.
+    auto rewriteBranch = [&](auto op, Block *dest) {
+      rewriter.setInsertionPointToEnd(op->getBlock());
+      rewriter.create<cf::BranchOp>(op.getLoc(), dest, op.getOperands());
+      rewriter.eraseOp(op);
+    };
+    for (Block *b = firstBlock; b != newLoopBlock; b = b->getNextNode())
+      for (auto &op : *b) {
+        if (auto contOp = dyn_cast<cudaq::cc::ContinueOp>(op)) {
+          rewriteBranch(contOp, newLoopBlock);
+          break;
+        } else if (auto brkOp = dyn_cast<cudaq::cc::BreakOp>(op)) {
+          rewriteBranch(brkOp, afterBlock);
+          break;
+        }
+      }
```

Comment:
Wouldn't calling `rewriter.eraseOp(op)` invalidate the iterators being used in line 88?

Reply:
Yep. That's why there is a `break`.